mesonbuild/coredata.py

```diff
@@ -421,8 +428,12 @@ def load(build_dir):
     if not isinstance(obj, CoreData):
         raise MesonException(load_fail_msg)
     if obj.version != version:
-        raise MesonException('Build directory has been generated with Meson version %s, which is incompatible with current version %s.\nPlease delete this build directory AND create a new one.' %
-                             (obj.version, version))
+        msg = 'Build directory has been generated with Meson version %s, which is incompatible with current version %s.\n'
+        if state_file_exists:
+            msg += 'Upgrade the build directory by invoking the "upgrade-builddir" target.'
```

Comment:
I'm pretty sure that upgrading by default will be something that will be wanted in the future (at least to support the case of automatic builds). But getting the manual version working first is OK.